Always HTTP 400 if marshmallow==3.0.0b13

Hi,

Just ran across this:

```
>               "Bad response: %s (not %s)\n%s", res_status, status, res)
E           webtest.app.AppError: Bad response: 400 Bad Request (not 200)
E           b'{"status": "error", "errors": [{"location": "cookies", "name": "cookies", "description": ["Unknown field."]}, {"location": "header", "name": "header", "description": ["Unknown field."]}, {"location": "method", "name": "method", "description": ["Unknown field."]}, {"location": "url", "name": "url", "description": ["Unknown field."]}, {"location": "path", "name": "path", "description": ["Unknown field."]}, {"location": "querystring", "name": "querystring", "description": ["Unknown field."]}]}'

../../../../.pyenv/versions/3.6.3/envs/p2server_py36/lib/python3.6/site-packages/webtest/app.py:686: AppError

```

The reason is marshmallow 3.0.0b13 https://github.com/marshmallow-code/marshmallow/pull/911

Perhaps `unknown = marshmallow.EXCLUDE` is now needed here:

https://github.com/Cornices/cornice/blob/fcda515b51a6fc781e08153b4b645733dcb1456a/cornice/validators/_marshmallow.py#L56-L58

The *catch* is that marshmallow wants to be more strict (than before) so *relaxing* it seems questionable.

Up to marshmallow 3.0.0b12, it all worked well. In 3.0.0b13, the `unknown=EXCLUDE` *fixes* the issue.

--- a/cornice/validators/_marshmallow.py
+++ b/cornice/validators/_marshmallow.py
@@ -38,6 +38,10 @@ def _generate_marshmallow_validator(location):
             :func:`cornice.validators.extract_cstruct`
         """
         import marshmallow
+        try:
+            from marshmallow.utils import EXCLUDE
+        except ImportError:
+            EXCLUDE = 'exclude'
 
         if schema is None:
             return
@@ -61,6 +65,7 @@ def _generate_marshmallow_validator(location):
         class Meta(object):
             strict = True
             ordered = True
+            unknown = EXCLUDE
 
         class RequestSchemaMeta(type):
             """

--- a/cornice/validators/_marshmallow.py
+++ b/cornice/validators/_marshmallow.py
@@ -117,6 +117,13 @@ def _message_normalizer(exc, no_field_name="_schema"):
     :return:
     """
     if isinstance(exc.messages, dict):
+        if '_schema' in exc.messages:
+            new_dict = {}
+            # if not dict expect a list of dicts and normalize it to just dict
+            if not hasattr(exc.messages['_schema'], 'keys'):
+                for item in exc.messages['_schema']:
+                    new_dict.update(item)
+                return {'_schema': new_dict}
         return exc.messages
     if len(exc.field_names) == 0:
         return {no_field_name: exc.messages}

--- a/tests/test_validation.py
+++ b/tests/test_validation.py
@@ -618,7 +618,7 @@ class TestServiceDefinitionMarshmallow(LoggingCatcher, TestCase):
         response = app.post_json('/m_newsletter?ref=2', params=content,
                                  status=400)
         self.assertEqual(response.status_code, 400)
-        self.assertEqual(response.json['errors'][0]['description'][0]['email'],
+        self.assertEqual(response.json['errors'][0]['description'],
                          'Invalid email length')
 
     def test_validated_path_content_from_schema(self):

--- a/tests/validationapp.py
+++ b/tests/validationapp.py
@@ -305,6 +305,10 @@ if COLANDER:
 
 try:
     import marshmallow
+    try:
+        from marshmallow.utils import EXCLUDE
+    except ImportError:
+        EXCLUDE = 'exclude'
     from cornice.validators import (
         marshmallow_validator,
         marshmallow_body_validator
@@ -327,9 +331,15 @@ if MARSHMALLOW:
 
 
     class MSignupSchema(marshmallow.Schema):
+        class Meta:
+            strict = True
+            unknown = EXCLUDE
         username = marshmallow.fields.String()
 
     class MSignupGroupSchema(marshmallow.Schema):
+        class Meta:
+            strict = True
+            unknown = EXCLUDE
         username = marshmallow.fields.String()
 
         def __init__(self, *args, **kwargs):
@@ -339,6 +349,9 @@ if MARSHMALLOW:
     import random
 
     class MNeedsContextSchema(marshmallow.Schema):
+        class Meta:
+            strict = True
+            unknown = EXCLUDE
         somefield = marshmallow.fields.Float(missing=lambda: random.random())
         csrf_secret = marshmallow.fields.String()
 
@@ -368,6 +381,9 @@ if MARSHMALLOW:
             raise Invalid(node, "The bar is not open.")
 
     class MBodySchema(marshmallow.Schema):
+        class Meta:
+            strict = True
+            unknown = EXCLUDE
         # foo and bar are required, baz is optional
         foo = marshmallow.fields.String()
         bar = SchemaNode(String(), validator=m_validate_bar)
@@ -377,9 +393,15 @@ if MARSHMALLOW:
         integers = marshmallow.fields.List(marshmallow.fields.Integer())
 
     class MQuery(marshmallow.Schema):
+        class Meta:
+            strict = True
+            unknown = EXCLUDE
         yeah = marshmallow.fields.String()
 
     class MRequestSchema(marshmallow.Schema):
+        class Meta:
+            strict = True
+            unknown = EXCLUDE
         body = marshmallow.fields.Nested(MBodySchema)
         querystring = marshmallow.fields.Nested(MQuery)
 
@@ -398,6 +420,10 @@ if MARSHMALLOW:
             return data
 
     class MQSSchema(marshmallow.Schema):
+        class Meta:
+            strict = True
+            unknown = EXCLUDE
+
         querystring = marshmallow.fields.Nested(MListQuerystringSequenced)
 
 
@@ -406,12 +432,22 @@ if MARSHMALLOW:
         return {"field": request.validated['querystring']['field']}
 
     class MNewsletterSchema(marshmallow.Schema):
+        class Meta:
+            strict = True
+            unknown = EXCLUDE
         email = marshmallow.fields.String(validate=marshmallow.validate.Email())
 
     class MRefererSchema(marshmallow.Schema):
+        class Meta:
+            strict = True
+            unknown = EXCLUDE
         ref = marshmallow.fields.Integer()
 
     class MNewsletterPayload(marshmallow.Schema):
+        class Meta:
+            strict = True
+            unknown = EXCLUDE
+
         body = marshmallow.fields.Nested(MNewsletterSchema)
         querystring = marshmallow.fields.Nested(MRefererSchema)
 
@@ -429,9 +465,15 @@ if MARSHMALLOW:
         return request.validated
 
     class MItemPathSchema(marshmallow.Schema):
+        class Meta:
+            strict = True
+            unknown = EXCLUDE
         item_id = marshmallow.fields.Integer(missing=None)
 
     class MItemSchema(marshmallow.Schema):
+        class Meta:
+            strict = True
+            unknown = EXCLUDE
         path = marshmallow.fields.Nested(MItemPathSchema)
 
 

--- a/tests/validationapp.py
+++ b/tests/validationapp.py
@@ -445,7 +445,6 @@ if MARSHMALLOW:
 
     class MNewsletterPayload(marshmallow.Schema):
         class Meta:
-            strict = True
             unknown = EXCLUDE
 
         body = marshmallow.fields.Nested(MNewsletterSchema)

